File: 0020-valid-parantheses/stack_approach.py
# Time complexity: O(N), Space complexity: O(N)
class Solution:
    def isValid(self, s: str) -> bool:
        
        size = len(s)
        
        if size % 2 != 0:
            return False
        
        stack = []
        
        for ch in s:
            if len(stack) == 0:
                stack.append(ch)
            else:
                st_top = stack[-1]
                if st_top == "(" and ch == ")":
                    stack.pop()
                elif st_top == "[" and ch == "]":
                    stack.pop()
                elif st_top == "{" and ch == "}":
                    stack.pop()
                else:
                    stack.append(ch)
                
        return len(stack) == 0

File: 0020-valid-parantheses/stack_approach.cpp
// Time complexity: O(N), Space complexity: O(N)
class Solution {
public:
    bool isValid(string s) {
        int len = s.length();
        
        if (len % 2 != 0) {
            return false;
        }
        
        stack<char> st;
        
        for (char ch : s) {
            if (st.empty()) {
                st.push(ch);
            }
            else {
                char stTop = st.top();
                if (stTop == '(' && ch == ')') {
                    st.pop();
                }
                else if (stTop == '[' && ch == ']') {
                    st.pop();
                }
                else if (stTop == '{' && ch == '}') {
                    st.pop();
                }
                else {
                    st.push(ch);
                }
            }
        }
        
        return st.empty();
    }
};